### PR TITLE
fix(bug): crash on takeover and info replication

### LINKED ISSUE #3280: Server crash while running replica takeover and info replication
The bug:
info replication first checks if the server is master, if not takes replica mutex and calls
dfly::Replica::GetInfo()
When running replica takeover we take the mutex finish the takeover reset the replica object and set is master to true.
With this flow when if we get the mutex after takeover is done but check is master before we will get a crash for accessing the replica object which was reset.

### PR DESCRIPTION
**Pull request by adiholden**
fixes #3280 
The bug:
info replication first checks if the server is master, if not takes replica mutex and calls
`dfly::Replica::GetInfo()`
When running replica takeover we take the mutex, finish the takeover, reset the replica object and set `is_master` to true.
With this flow when if we get the mutex after takeover is done but check is master before we will get a crash for accessing the replica object which was reset.
The fix:
take the mutex before checking is_master

### CODE REVIEW

**Review by chakaz — DISMISSED** [review]

**Comment by romange** [comment]
I am asking myself then if `etl.is_master` is working at all.

File: src/server/server_family.cc [hunk 1]
```diff
@@ -2258,6 +2258,7 @@ void ServerFamily::Info(CmdArgList args, ConnectionContext* cntx) {
   }
 
   if (should_enter("REPLICATION")) {
+    unique_lock lk(replicaof_mu_);
```
Comment (romange) [inline src/server/server_family.cc:2261]: if you take `replicaof_mu_`, you should not rely here on `etl.is_master` and you can check `replica_` instead.
also, a comment saying that `etl.is_master` is eventually consistent and does not fit here would explain why you access `replica_` directly.

File: src/server/server_family.cc [hunk 2]
```diff
@@ -2737,6 +2734,7 @@ void ServerFamily::ReplConf(CmdArgList args, ConnectionContext* cntx) {
 
 void ServerFamily::Role(CmdArgList args, ConnectionContext* cntx) {
   auto* rb = static_cast<RedisReplyBuilder*>(cntx->reply_builder());
+  unique_lock lk(replicaof_mu_);
```
Comment (romange) [inline src/server/server_family.cc:2737]: same here, you can access `replica_` instead of etl.master

**Review by romange — COMMENTED** [review]
good catch

**Review by romange — APPROVED** [review]
Thanks!

**Pull request merged**